### DISCUSSION TITLE
Auth Session Starts Authenticated

### DISCUSSION THREAD

**Discussion by boopxyz** [nextauthjs/next-auth#12367] [Help] [answered]
So I've got a problem where my session is starting out valid and authenticated which leads to my things hidden to the unauth'ed client being shown briefly.
![image](https://github.com/user-attachments/assets/8131c549-7c3e-4ca1-9d7d-50d8d1735001)

**Comment by nic-vo** [comment]
@boopxyz Are you using the next-auth/react components?

**Reply by boopxyz** [reply]
What do you mean by that?

**Reply by nic-vo** [reply]
Are you checking auth client-side with the `<SessionProvider />` and `useSession` hooks that next-auth provides? I'm assuming yes since your screencap looks like a browser console

**Reply by boopxyz** [reply]
Yes that is what I'm using

**Reply by nic-vo** [reply]
How are you checking the `session` returned by the `useSession` hook in your React component? Typically it's a good idea to return any protected JSX only if `status === "authenticated"`. And I believe that typically, the client-side components don't even begin to check auth until they're mounted (the auth check happens in a `useEffect`), which means `status` is typically `"loading"` on initial navigation

**Reply by boopxyz** [reply]
Well the thing is the useSession's status starts as authenticated which shouldn't be because it should check it first, not have that be its initial stage.

**Reply by nic-vo** [reply]
Hmm yeah you're right.

So these console logs are from initial navigation to a page / route? Without the user touching any login / log out button?

Where is your `<SessionProvider />` in the component tree? Root layout?

**Reply by boopxyz** [reply]
> Hmm yeah you're right.
> 
> So these console logs are from initial navigation to a page / route? Without the user touching any login / log out button?
> 
> Where is your `<SessionProvider />` in the component tree? Root layout?

> [!IMPORTANT]
> Using the latest version of Next.js v14 (not v15) with the app router 

Correct, it's from the start of loading a page. My SessionProvider is in a client component (given I'm using the app router) which is in the layout

```javascript
<Providers session={getSession()} />
```
which equals
```javascript
<SessionProvider session={session}>{children}</SessionProvider>
```

**Reply by nic-vo** [reply]
Oh okay I think I might know what's going on. Don't pass the session into the `SessionProvider` as you have done. Also make sure you use TypeScript

1) that `getSession()` function [is async](https://next-auth.js.org/v3/getting-started/client#getsession), at least if it's the one imported from `next-auth/react`. You have to await it higher up in the function body and pass the resolved session as the Provider component `session` prop. 

2) `SessionProvider` [doesn't have a really explicit check for a `Session`](https://github.com/nextauthjs/next-auth/blob/7969dd59e708038e66ca421e9060f501ff18f739/packages/next-auth/src/react.tsx#L385C1-L389C56). It assumes that the `session` prop is actually a `Session`, but technically a non-`Session` prop can be passed (if you're not using TypeScript; compiler would throw a fit otherwise). 

3) `SessionProvider` will automatically check session for you on mount in a `useEffect`. You don't need to explicitly pass it a session for it to function properly. It'll have the expected `'loading'` -> `'unauthenticated'` | `'authenticated'` behavior

**Reply by LDrex1** [reply, marked as answer]
> > Hmm yeah you're right.
> > So these console logs are from initial navigation to a page / route? Without the user touching any login / log out button?
> > Where is your `<SessionProvider />` in the component tree? Root layout?
> 
> Important
> 
> Using the latest version of Next.js v14 (not v15) with the app router
> 
> Correct, it's from the start of loading a page. My SessionProvider is in a client component (given I'm using the app router) which is in the layout
> 
> ```js
> <Providers session={getSession()} />
> ```
> 
> which equals
> 
> ```js
> <SessionProvider session={session}>{children}</SessionProvider>
> ```

This doesn't look right.

Why don't you pass it this way from your async layout component or whereever 
```js
const session = await getServerSession(authOptions)
// or
const session  = await getSession()
```

It returns a promise so you have to await the result. 
if in the child components you are doing something like
```js
session &&
```

It might consider session truthy if you didn't await it because getSession() returns a promise.
I'd advice you to read the docs before going further.
[getSession] (https://next-auth.js.org/getting-started/client#getsession)

**Comment by LDrex1** [comment]
I think the best way would be to use the middleware to match your protected routes and redirect to your login url based on if the token from next auth exists or not. That way, if the user isn't authenticated, it will redirect to the login page from the middleware before it even gets to the protected routes. 
Note: easiest way is to use the jwt strategy which you can set in the authOptions. If not, you'll have to handle the authentication using cookies in the middleware.

**Reply by boopxyz** [reply]
Well I need the login/logout button to be visible on all pages, including those with no logins. The issue is the status given by useSession is never loading, it's starting out as authenticated which shouldn't be the case.

**Comment by boopxyz** [comment]
Still having problems with this. Any help with be great!
